### PR TITLE
172921509 unauthorized e2e tests

### PR DESCRIPTION
**Pull request by guo-dna**
- Adds a few test cases for unauthorized 401s and /actuator/health
- Supply a www-authenticate header on 401 response from Google
This is a pretty optional change tbh. The existing test cases were already pretty well done.

### CODE REVIEW

File: src/main/java/com/dnastack/search/sheets/GlobalControllerExceptionHandler.java [hunk 1]
```diff
@@ -35,6 +35,11 @@
     @ExceptionHandler(GoogleJsonResponseException.class)
     public ResponseEntity<SearchErrorResponse> handle(GoogleJsonResponseException ex) {
         log.info("Returning " + ex.getStatusCode(), ex);
+        if (ex.getStatusCode() == 401) {
```
Comment (guo-dna) [inline src/main/java/com/dnastack/search/sheets/GlobalControllerExceptionHandler.java:38]: Should be int primitive comparison, not object. So I used ==...

Comment (FizzDNA) [inline src/main/java/com/dnastack/search/sheets/GlobalControllerExceptionHandler.java:38]: Do 401 statuses need to be special-cased like this, or could the authentication headers be included on any error of this type? If the 401 status does need to be special-cased, do we need to do the same for 403?

File: e2e-tests/src/main/java/com/dnastack/search/sheets/UnauthorizedRespE2eTest.java [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+        given()
+            .log().uri()
+            .header("Authorization", "Bearer " + "A_malformed_bearer_token")
+        .when()
+            .get(endpoint)
+        .then()
+            .statusCode(401)
+            // Attempt at future-proofing since the header is from Google. Avoid exact match of full string.
+            .header("WWW-Authenticate", containsStringIgnoringCase("Bearer"))
+            .header("WWW-Authenticate", containsStringIgnoringCase("invalid_token"));
```
Comment (FizzDNA) [inline e2e-tests/src/main/java/com/dnastack/search/sheets/UnauthorizedRespE2eTest.java:51]: This is probably OK in terms of keying on strings in the response, as I believe that the "invalid_token" string is a standardized response for this error: https://tools.ietf.org/html/rfc6750#section-6.2.2

**Review by FizzDNA — CHANGES_REQUESTED** [review]

File: e2e-tests/src/main/java/com/dnastack/search/sheets/UnauthorizedRespE2eTest.java [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+    }
+
+    private void getWithoutAuthHeaderAndCheckFor400(String endpoint) {
+        given()
+            .log().uri()
+        .when()
+            .get(endpoint)
+        .then()
+            .statusCode(400)
+            .body("message", equalTo("Missing request header 'authorization' for method parameter of type String"));
```
[the remark this answers is not in the document]
Comment (guo-dna) [reply e2e-tests/src/main/java/com/dnastack/search/sheets/UnauthorizedRespE2eTest.java]: I agree with you. It's brittle to capture the 400 response body, especially given that this seems to be generated by Spring. I think checking for just 400 should be sufficient and I pushed a change removing that line.

File: src/main/java/com/dnastack/search/sheets/GlobalControllerExceptionHandler.java (hunk 1 above)
Comment (guo-dna) [reply src/main/java/com/dnastack/search/sheets/GlobalControllerExceptionHandler.java:38]: Maybe we can discuss a bit. The (minor) "issue" here was there was no `www-authenticate` header in a 401 response. I didn't see an easy way to forward the entire header without a cast or a converter class so I gave it a special case. Is a 403 required to have the "www-authenticate" header?
Screen Shot 2020-09-24 at 3 07 05 PM
It's fair to claim that its a band-aid fix, and I'd be happy to discuss if this was even a true "issue" to begin with.

**Review by FizzDNA — APPROVED** [review]
Tests and reasoning are sound!

LGTM

**Pull request merged**